joey/layers_v2.py

```diff
@@ -391,3 +393,357 @@ def __init__(self, kernel_size, input_size,
+            input_d = input_size[-self._dims+i] + 2 * padding[i]
+            if (input_d - kernel_size[-self._dims+i]) % stride[i] != 0:
+                raise Exception("Stride " + str(stride) + " is not "
+                                "compatible with feature map, kernel and "
+                                "padding sizes. If you want to proceed "
+                                "anyway, set strict_stride_check=False "
+                                "when instantiating this object")
+
+    def _allocate(self, kernel_size, input_size, name_allocator_func,
+                  dim_allocator_func):
+
+        dimensions = ['dbatch', 'dchannel']
+        result_shape = []
+        input_size = list(input_size)
+        # generating  in the order depth, height, width ,
+        # hence arr[-3], arr[-2] and so on
+        for i in range(0, self._dims):
+            result_d = (input_size[(-self._dims+i)] -
+                        kernel_size[(-self._dims+i)] +
+                        2 * self._padding[i])//self._stride[i] + 1
+            result_shape.append(result_d)
+            input_size[(-self._dims+i)] += 2 * self._padding[i]
+            dimensions.append(
+                'd_'+self._dim_dict.get(self._dims-i, self._dims-i))
+
+        result_shape = (*input_size[0:2], *result_shape)
+        # input data function
+        input_dimensions = [SpaceDimension("Input_"+x) for x in dimensions]
+
+        input_func = Function(name="Input_F", shape=(input_size),
+                              dimensions=input_dimensions,
+                              space_order=0, dtype=np.float64)
+
+        # Result for pool layer
+        result_dimensions = [SpaceDimension("Result_"+x) for x in dimensions]
+
+        result_func = Function(name="Result_F", shape=result_shape,
+                               dimensions=result_dimensions, space_order=0,
+                               dtype=np.float64)
+
+        output_grad_dimensions = [SpaceDimension(
+            "output_grad"+x) for x in dimensions]
+
+        output_grad = Function(name="outgrad_%s" % name_allocator_func(
+        ), shape=result_shape, dimensions=output_grad_dimensions,
+            space_order=0, dtype=np.float64)
+
+        self._set_padding_result_values(
+            input_func, result_func, value=np.finfo(np.float32).min)
+
+        return (None, input_func, result_func, None, None, output_grad, None)
+
+    @property
+    def stride(self):
+        """Stride of the layer."""
+        return self._stride
+
+    @property
+    def kernel_size(self):
+        """The kernel size of the layer."""
+        return self._kernel_size
+
+    def execute(self, input_data):
+        '''execute implementation'''
+        indices = [slice(0, self._I.shape[0], 1),
+                   slice(0, self._I.shape[1], 1)]
+        [indices.append(slice(self._padding[i],
+                              self._I.data.shape[2+i]-self._padding[i], 1))
+            for i in range(self._dims)]
```

Comment:
The previous few lines are really hard to read. If a list comprehension spans three lines, it might be better written as a for loop.